### PR TITLE
Fix exhaustivity check of `classOrArrayType`

### PR DESCRIPTION
**Pull request by liufengyun**
Previous the usage of extractor produces an unexhaustive warning. Anyway, the previous code seems problematic to me, as `this` can never be `null`, the check in the extractor is useless.

### CODE REVIEW

**Comment by smarter** [comment]
> Anyway, the previous code seems problematic to me, as this can never be null, the check in the extractor is useless.

I agree. Could you change the extractor to remove the null check and return `Some[String]` instead of changing the case?

File: src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala [hunk 1]
```diff
@@ -754,9 +754,7 @@ abstract class BTypes {
     /**
      * Pattern matching on a ClassBType extracts the `internalName` of the class.
      */
-    def unapply(c: ClassBType): Option[String] =
-      if (c == null) None
-      else Some(c.internalName)
+    def unapply(c: ClassBType): Some[String] = Some(c.internalName)
```
Comment (liufengyun) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:757]: I just checked in source code, did not find places that the null check is useful -- all usage of it is in the context of matching against `this`. So I assume this change is safe.

Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:757]: why was this changed? Backend used to use `null` as a value.

**Review by DarkDimius — COMMENTED** [review]
> as this can never be null

Which "this" do you mean?

**Comment by liufengyun** [comment]
@DarkDimius I searched the code, the usage are in following locations:

```Scala
63:      case ClassBType(internalName) => "L" + internalName + ";"
239:      case ClassBType(internalName) => asm.Type.getObjectType(internalName) // see (*) above
338:      case ClassBType(internalName) => internalName
```

All of them have the structure:

```Scala
this match {
  case ClassBType(internalName) => ...
}
```

**Review by smarter — APPROVED** [review]
LGTM if the first commit is removed since it shouldn't be necessary to get rid of the warning. The CI is failing because ScalacBackendInterface has been broken since the semantic name PR, this doesn't have to be fixed in this PR.

**Comment by smarter** [comment]
Merging now, @liufengyun : please make a PR against dotty to update the backend to this commit.

**Comment by liufengyun** [comment]
Thanks a lot @smarter !
